scripts/deployed_general_agent_viewer.py

```diff
@@ -0,0 +1,119 @@
+"""
+PYTHONPATH=. streamlit run deployed_agent_viewer.py
+
+Tip: if you specify PYTHONPATH=., streamlit will watch for the changes in all files, instead of just this one.
+"""
+
+
+# Imports using asyncio (in this case mech_client) cause issues with Streamlit
+from prediction_market_agent.tools.streamlit_utils import (  # isort:skip
+    display_chat_history,
+    streamlit_asyncio_event_loop_hack,
+)
+
+streamlit_asyncio_event_loop_hack()
+
+# Fix "Your system has an unsupported version of sqlite3. Chroma requires sqlite3 >= 3.35.0" error
+from prediction_market_agent.utils import patch_sqlite3  # isort:skip
+
+patch_sqlite3()
+
+
+from datetime import datetime
+
+import streamlit as st
+from prediction_market_agent_tooling.markets.markets import MarketType
+
+from prediction_market_agent.agents.microchain_agent.memory import (
+    DatedChatHistory,
+    LongTermMemory,
+)
+from prediction_market_agent.agents.microchain_agent.utils import get_total_asset_value
+from prediction_market_agent.agents.utils import AgentIdentifier
+from prediction_market_agent.tools.streamlit_utils import check_required_api_keys
+from prediction_market_agent.utils import APIKeys
+
+
+class DeployedGeneralAgentKeys(APIKeys):
+    START_TIME: str
+    STARTING_BALANCE: float
+
+    @property
+    def start_time(self) -> datetime:
+        return datetime.fromisoformat(self.START_TIME)
+
+    @property
+    def starting_balance(self) -> float:
+        return self.STARTING_BALANCE
+
+
+MARKET_TYPE = MarketType.OMEN
+currency = MARKET_TYPE.market_class.currency
+
+st.set_page_config(
+    layout="wide",
+    page_title="Gnosis AI: Deployed Prediction Market Trader Agent",
+    page_icon=":owl:",
+)
+st.title("Deployed Trader Agent Viewer")
+
+check_required_api_keys(["BET_FROM_PRIVATE_KEY"])
+keys = DeployedGeneralAgentKeys()
+starting_balance = keys.starting_balance
```

Comment:
I did look at doing that, but from what I could see, finding `block_from_prev_tx` was a bit of a pain (e.g. `def find_block_by_timestamp(timestamp: datetime)` would need binary search to find the block id).

Reply:
I think it's a bit of work, so would leave it to you to either do here or create a new issue. What I had in mind would be:
-> Implement a query using the API from Gnosis blockscout (https://gnosis.blockscout.com/api-docs) (https://gnosis.blockscout.com/api/v2/addresses/0x6063a8de125d0982fcd7b41754262d94cac8ff14/transactions?filter=from) so that you have all transactions from a given address
-> Get the first tx from the above sequence, extract block, call `get_balance` with a fixed block.

Reply:
Feel free to close this if you prefer to create a separate issue

Reply:
Don't think it's a great idea to manually update things every time we refund.
The way I understand it, ROI divides earnings by the cost of the investment (https://www.investopedia.com/terms/r/returnoninvestment.asp#:~:text=How%20Do%20You%20Calculate%20Return,when%20expressed%20as%20a%20percentage.). If further deposits were made, ofc ROI changes.

What I am concretely suggesting here is to make these quantities (ROI, starting balance) dynamic to avoid repetitive, error-prone manual work.

Reply:
Okay, but how will you do it? Scan all the transactions of the agent and make sure that #finance people always send xDai from the same address so that we can automatically subtract it?

Reply:
In my head,
`ROI = earnings/capital`. This can then be averaged during a rolling 30-day period to get the yearly value.
-> Earnings are comprised of `redeemWinnings` transactions. Let's assume agent received 1 xDAI on day 1 of month and 2 xDAI on day 15 of month.
-> Capital is the average balance of the agent over the entire month. The balance is constantly changing, due to top-ups (as @kongzii said), and due to trading itself. Hence we can either a. fetch all transactions by the addresss that are bets and subtract the balance (not so nice) or b. sample N blocks during the month and average the balance.

Sounds complicated but would give us a precise measure of ROI.

Reply:
> a. fetch all transactions by the addresss that are bets and subtract the balance (not so nice)

How do we know from a transaction that it comes from a bet?

> Let's assume agent received 1 xDAI on day 1 of month and 2 xDAI on day 15 of month.

> b. sample N blocks during the month and average the balance.

If we deposit 100 xDai on 10 of the month. How will this filter it out?

Reply:
> How do we know from a transaction that it comes from a bet?
You can fetch all txs from a given address (using the Blockscout API), and then you select only the tx that contain an event (`FPMMSell` or `FPMMBuy`). This is called Bloom filter (https://web3py.readthedocs.io/en/v5/filters.html)

> If we deposit 100 xDai on 10 of the month. How will this filter it out?
The idea is to have an average of the balance. You can sample every block during the 30 day period (lots of RPC calls), or you can sample n/10, n/20, etc and get a less precise estimate. 
One could also interpret the transactions to find deposits, but agree this would be more cumbersome, hence the sampling suggestion.

Reply:
Sorry, I still don't see how (2) will help to fix this. If we deposit 100 xDai to the wallet on the 10th, then it will be there every day after that until the agent spends it entirely. So it will affect averages from 10th to the end of the month, and then next one, etc.

For (1), sounds really complicated and more error-prone then updating it manually every few months, when we add funds.

But! If you guys can show me wrong in the code, I'll be happy to withdraw my worries 😄

Reply:
(1)
-> It involves one API call (see [comment](https://github.com/gnosis/prediction-market-agent/pull/266#discussion_r1651474131)) and then a one-time implementation of a filter for txs containing one of 2 events. It definitely involves work but it has to be done once vs every time we update a balance manually, hence would strongly advocate for this.

(2)
Let's assume we have these earnings
- day 1: 1 xDAI (prev balance 10)
- day 10: 0 xDAI (prev balance 11, new deposit of 10 xDAI arrived, total balance 21)
- day 20: 1 xDAI (balance 22)
Following the above, we have 3 xDAI of `total_earnings` and an average balance of
`10 days of balance  * 11 / 30` + `10 days of balance * 21 / 30` + `10 days of balance * 22/30` = **18**
So we would have 2 xDAI / 18 = 11% monthly return.
An alternate logic (which is what I'm proposing) is to fetch the balance for each day and average those, so instead of having 10 day periods (as per above) we would have `balance_day1/30` + `balance_day2/30` and so on. Does this make sense to you?

Reply:
My thought is that we only fund the agent once, at the start, then if it runs out of money, analyse its chat history, see why it failed, try to improve the agent, and deploy a new agent. So no top-up funding.

That also makes the ROI calculation simple (as it currently is in this PR), and doesn't introduce this extra complexity 😄.

Agree though that determining the starting balance programatically is still a better solution. Separate issue for that here https://github.com/gnosis/prediction-market-agent/issues/273

Reply:
> Does this make sense to you?

Still doesn't, but Evan's idea how it will work solved it anyway 😄